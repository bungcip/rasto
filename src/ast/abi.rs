//! The `ast` module contains the definitions for the Abstract Syntax Tree (AST) nodes
//! that represent Rust ABIs.

use crate::pretty_printer::{PrettyPrinter, Printer};
use std::fmt;

/// Represents a Rust ABI (Application Binary Interface).
///
/// This enum is used to specify the calling convention for functions,
/// especially in the context of `extern` blocks and function pointers.
///
/// # Examples
///
/// ```rust
/// use rasto::ast::Abi;
///
/// // A named ABI, like "C"
/// let c_abi = Abi::Named("C".to_string());
///
/// // A named ABI, like "system"
/// let system_abi = Abi::Named("system".to_string());
/// ```
#[derive(Debug, Clone, PartialEq)]
pub enum Abi {
    /// Represents a named ABI, specified as a string literal.
    ///
    /// For example, `"C"`, `"system"`, or `"Rust"`.
    Named(String),
}

impl PrettyPrinter for Abi {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Abi::Named(name) => {
                printer.string("\"");
                printer.string(name);
                printer.string("\"");
            }
        }
        Ok(())
    }
}
