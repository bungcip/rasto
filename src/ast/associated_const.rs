//! An associated constant in a trait or impl.

use crate::ast::expressions::Expr;
use crate::ast::ident::Ident;
use crate::ast::metadata::Md;
use crate::ast::types::Type;

/// Represents an associated constant in a trait or implementation.
///
/// An associated constant is a constant value that is associated with a specific
/// type. It is defined within a trait or an `impl` block.
///
/// # Examples
///
/// In a trait definition:
///
/// ```rust
/// trait MyTrait {
///     const MAX: u16;
/// }
/// ```
///
/// In an `impl` block:
///
/// ```rust
/// trait MyTrait {
///     const MAX: u16;
/// }
/// struct MyType;
///
/// impl MyTrait for MyType {
///     const MAX: u16 = 123;
/// }
/// ```
#[derive(Debug, Clone, PartialEq)]
pub struct AssociatedConst {
    /// The name of the associated constant.
    pub ident: Ident,
    /// The type of the associated constant.
    pub ty: Type,
    /// The optional expression that provides the value of the constant.
    ///
    /// This is `None` in a trait definition where the value is not specified,
    /// and `Some` in an `impl` block where the value is provided.
    pub expr: Option<Box<Expr>>,
    /// Metadata associated with the constant, such as attributes and comments.
    pub md: Option<Box<Md>>,
}
