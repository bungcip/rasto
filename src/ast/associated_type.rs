use crate::ast::generics::GenericParams;
use crate::ast::metadata::Md;
use crate::ast::types::Type;
use thin_vec::ThinVec;

/// An associated type in a trait.
#[derive(Debug, Clone, PartialEq)]
pub struct AssociatedType {
    /// The name of the associated type.
    pub ident: String,
    /// The generic parameters of the associated type.
    pub generics: GenericParams,
    /// The bounds on the associated type.
    pub bounds: ThinVec<Type>,
    /// The default type for the associated type.
    pub default: Option<Type>,
    /// Metadata about the associated type.
    pub md: Option<Box<Md>>,
}
