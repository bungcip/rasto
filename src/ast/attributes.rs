//! Defines the AST nodes for attributes.
//!
//! Attributes are metadata that can be attached to various items in Rust code. They are
//! enclosed in `#[...]` for outer attributes and `#![...]` for inner attributes.

use crate::ast::literals::Lit;
use crate::pretty_printer::{PrettyPrinter, Printer};
use std::fmt;
use thin_vec::ThinVec;

/// An attribute, such as `#[repr(C)]` or `#![allow(dead_code)]`.
#[derive(Debug, Clone, PartialEq)]
pub enum Attribute {
    /// An inner attribute, such as `#![allow(dead_code)]`.
    Inner(Meta),
    /// An outer attribute, such as `#[repr(C)]`.
    Outer(Meta),
}

/// The meta item within an attribute.
///
/// For example, in `#[repr(C)]`, the meta item is `repr(C)`.
#[derive(Debug, Clone, PartialEq)]
pub enum Meta {
    /// A meta list, such as `repr(C)`.
    List(MetaList),
    /// A meta path, such as `test`.
    Path(String),
    /// A meta name-value pair, such as `key = "value"`.
    NameValue(MetaNameValue),
}

/// A meta list, such as `repr(C)`.
#[derive(Debug, Clone, PartialEq)]
pub struct MetaList {
    /// The path of the meta list, e.g., `repr`.
    pub path: String,
    /// The meta items within the list, e.g., `C`.
    pub metas: ThinVec<Meta>,
}

/// A meta name-value pair, such as `key = "value"`.
#[derive(Debug, Clone, PartialEq)]
pub struct MetaNameValue {
    /// The path of the meta name-value pair, e.g., `key`.
    pub path: String,
    /// The value of the meta name-value pair, e.g., `"value"`.
    pub value: Lit,
}

impl PrettyPrinter for Attribute {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Attribute::Inner(meta) => {
                printer.string("#![");
                meta.pretty_print(printer)?;
                printer.string("]");
            }
            Attribute::Outer(meta) => {
                printer.string("#[");
                meta.pretty_print(printer)?;
                printer.string("]");
            }
        }
        Ok(())
    }
}

impl PrettyPrinter for Meta {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Meta::List(list) => list.pretty_print(printer),
            Meta::Path(path) => {
                printer.string(path);
                Ok(())
            }
            Meta::NameValue(name_value) => name_value.pretty_print(printer),
        }
    }
}

impl PrettyPrinter for MetaList {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        printer.string(&self.path);
        printer.string("(");
        for (i, meta) in self.metas.iter().enumerate() {
            if i > 0 {
                printer.string(", ");
            }
            meta.pretty_print(printer)?;
        }
        printer.string(")");
        Ok(())
    }
}

impl PrettyPrinter for MetaNameValue {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        printer.string(&self.path);
        printer.string(" = ");
        self.value.pretty_print(printer)
    }
}
