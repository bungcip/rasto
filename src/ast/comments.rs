//! Defines the representation of comments in the AST.

use compact_str::CompactString;

/// Represents a comment in the source code.
///
/// Comments can be either line comments or block comments.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Comment {
    /// A line comment, starting with `//`.
    ///
    /// The string contains the content of the comment, without the `//`.
    ///
    /// # Example
    ///
    /// ```text
    /// // This is a line comment.
    /// ```
    Line(CompactString),
    /// A block comment, enclosed in `/* ... */`.
    ///
    /// The string contains the content of the comment, without the `/*` and `*/`.
    ///
    /// # Example
    ///
    /// ```text
    /// /* This is a block comment. */
    /// ```
    Block(CompactString),
    /// A doc comment, starting with `///`.
    ///
    /// The string contains the content of the comment, without the `///`.
    ///
    /// # Example
    ///
    /// ```text
    /// /// This is a doc comment.
    /// ```
    Doc(CompactString),
}
