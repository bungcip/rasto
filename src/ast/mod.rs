//! The `ast` module contains the definitions for the Abstract Syntax Tree (AST) nodes.
//!
//! This module defines the data structures that represent the syntactic structure of Rust code.
//! It includes nodes for expressions, statements, items, literals, and more. The AST is used
//! by the pretty-printer to generate formatted Rust code.

#[macro_use]
mod macros;

/// Defines the AST node for an `extern` block.
pub mod abi;
/// Defines the AST node for an associated const in a trait.
pub mod associated_const;
/// Defines the AST node for an associated type in a trait.
pub mod associated_type;
/// Defines the AST nodes for attributes.
pub mod attributes;
/// Defines the AST nodes for comments.
pub mod comments;
/// Defines the AST nodes for expressions.
pub mod expressions;
/// Defines the AST node for a file.
pub mod file;
/// Defines the AST nodes for generics.
pub mod generics;
/// Defines the AST node for an `asm!` expression.
pub mod item_asm;
/// Defines the AST node for a `const`, `static`, or `type` item.
pub mod item_def;
/// Defines the AST node for an enum.
pub mod item_enum;
/// Defines the AST node for an `extern` block.
pub mod item_extern_block;
/// Defines the AST node for an `extern crate` item.
pub mod item_extern_crate;
/// Defines the AST node for a function.
pub mod item_fn;
/// Defines the AST node for a foreign module.
pub mod item_foreign_mod;
/// Defines the AST node for an `impl` item.
pub mod item_impl;
/// Defines the AST node for a macro definition.
pub mod item_macro;
/// Defines the AST node for a module.
pub mod item_mod;
/// Defines the AST node for a static item.
pub mod item_static;
/// Defines the AST node for a struct.
pub mod item_struct;
/// Defines the AST node for a trait.
pub mod item_trait;
/// Defines the AST node for a trait alias.
pub mod item_trait_alias;
/// Defines the AST node for a `union` item.
pub mod item_union;
/// Defines the AST node for a `use` item.
pub mod item_use;
/// Defines the AST nodes for items.
pub mod items;
/// Defines the AST nodes for keywords.
pub mod keyword;
/// Defines the AST nodes for literals.
pub mod literals;
/// Defines the AST nodes for metadata.
pub mod metadata;
/// Defines the AST nodes for patterns.
pub mod patterns;
/// Defines the AST nodes for statements.
pub mod statements;
/// Defines the AST nodes for tokens.
pub mod tokens;
/// Defines the AST nodes for types.
pub mod types;
/// Defines the AST nodes for visibility.
pub mod visibility;
/// Defines the AST nodes for `where` clauses.
pub mod where_clause;

pub use abi::*;
pub use associated_const::*;
pub use associated_type::*;
pub use attributes::*;
pub use comments::*;
pub use expressions::*;
pub use file::*;
pub use generics::*;
pub use item_asm::*;
pub use item_def::*;
pub use item_enum::*;
pub use item_extern_block::*;
pub use item_extern_crate::*;
pub use item_fn::*;
pub use item_foreign_mod::*;
pub use item_impl::*;
pub use item_macro::*;
pub use item_mod::*;
pub use item_static::*;
pub use item_struct::*;
pub use item_trait::*;
pub use item_trait_alias::*;
pub use item_union::*;
pub use item_use::*;
pub use literals::*;
pub use metadata::*;
pub use patterns::*;
pub use statements::*;
pub use tokens::*;
pub use types::*;
pub use visibility::*;
pub use where_clause::*;
