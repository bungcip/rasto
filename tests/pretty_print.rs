use rasto::ast::*;
use rasto::builder::{
    block, comment, enum_def, expr, field_value, file, fn_def, impl_block, pat, stmt, struct_def,
    trait_def, trait_item_fn,
};
use rasto::pretty;
use thin_vec::thin_vec;

mod patterns;

#[test]
fn test_file() {
    let ast = file()
        .item(
            struct_def("Foo")
                .leading_comment(comment().line(" A simple struct."))
                .field("field1", "i32")
                .field("field2", "String")
                .build(),
        )
        .item(
            fn_def("foo")
                .input(pat().ident("a", false))
                .output("i32")
                .statement(expr().lit(42))
                .build(),
        )
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_macro_call_with_brackets() {
    let ast = fn_def("foo")
        .statement(expr().macro_call(
            "vec",
            Delimiter::Bracket,
            thin_vec![
                TokenTree::Literal(0.into()),
                TokenTree::Punct(Punct {
                    ch: ';',
                    spacing: Spacing::Alone,
                }),
                TokenTree::Literal(256.into()),
            ],
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_pretty_print_doc_comment() {
    let a = file()
        .item(
            fn_def("foo")
                .leading_comment(comment().doc(" This is a doc comment."))
                .build(),
        )
        .build();

    insta::assert_snapshot!(pretty(&a), @r"
    /// This is a doc comment.
    fn foo() {}
    ");
}

#[test]
fn test_block_single_comment() {
    let single = comment().block("Block comment with single line");
    insta::assert_snapshot!(pretty(&single));
}

#[test]
fn test_block_multiline_comment() {
    let single =
        comment().block("Block comment with multi line 1\nBlock comment with multi line 2");
    insta::assert_snapshot!(pretty(&single));
}

#[test]
fn test_fn() {
    let ast = fn_def("foo")
        .input(pat().ident("a", false))
        .output("i32")
        .block(
            block()
                .leading_comment(comment().block(" Block comment with single line "))
                .statement(
                    stmt()
                        .local(pat().ident("hello", false))
                        .expr(expr().lit("world")),
                )
                .statement(expr().lit(42))
                .has_trailing_semicolon(false),
        )
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_block_with_comments() {
    let ast = block()
        .leading_comment(comment().line(" leading comment"))
        .statement(expr().lit(42))
        .trailing_comment(comment().line(" trailing comment"))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn unary_expression() {
    let ast = expr().unary(UnOp::Neg, expr().lit(42));
    insta::assert_snapshot!(pretty(&ast), @"-42");
}

#[test]
fn unary_expression_not() {
    let ast = expr().unary(UnOp::Not, expr().lit(true));
    insta::assert_snapshot!(pretty(&ast), @"!true");
}

#[test]
fn test_expr_array() {
    let ast = expr().array([expr().lit(1), expr().lit(2)]);
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_async() {
    let ast = expr().async_block(block().statement(expr().lit(1)).build());
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_await() {
    let ast = expr().await_expr(expr().lit("future"));
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_break() {
    let ast = expr().break_expr();
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_call() {
    let ast = expr().call(expr().lit("foo"), vec![expr().lit(1), expr().lit(2)]);
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_cast() {
    let ast = expr().cast(expr().lit("x"), "u32");
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_closure() {
    let ast = expr().closure(
        vec![pat().ident("a", false), pat().ident("b", false)],
        expr().lit(1),
    );
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_const() {
    let ast = expr().const_block(block().statement(expr().lit(1)).build());
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_continue() {
    let ast = expr().continue_expr();
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_field() {
    let ast = expr().field(expr().lit("stru"), "field");
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_index() {
    let ast = expr().index(expr().lit("arr"), expr().lit(0));
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_match() {
    let ast = expr().match_expr(
        expr().lit("x"),
        vec![
            Arm {
                pat: pat().tuple(vec![pat().ident("Some", false), pat().ident("y", false)]),
                guard: None,
                body: Box::new(expr().lit(1)),
            },
            Arm {
                pat: pat().ident("None", false),
                guard: Some(Box::new(expr().lit(true))),
                body: Box::new(expr().lit(2)),
            },
        ],
    );
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_method_call() {
    let ast = expr().method_call(
        expr().lit("obj"),
        "method",
        vec![expr().lit(1), expr().lit(2)],
    );
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_paren() {
    let ast = expr().paren(expr().lit(1));
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_range() {
    let ast = expr().range(
        Some(expr().lit(1)),
        RangeLimits::HalfOpen,
        Some(expr().lit(5)),
    );
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_reference() {
    let ast = expr().reference(true, expr().lit("x"));
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_return() {
    let ast = expr().return_expr(Some(expr().lit(1)));
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_struct() {
    let ast = expr().struct_expr(
        "Foo",
        vec![
            field_value("a", expr().lit(1)),
            field_value("b", expr().lit(2)),
        ],
    );
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_tuple() {
    let ast = expr().tuple(vec![expr().lit(1), expr().lit(2)]);
    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_long_enum() {
    let ast = enum_def("MyLongLongLongLongLongEnum")
        .variant("AVeryLongVariantNameThatShouldCauseALineBreak")
        .variant("AnotherVeryLongVariantNameThatShouldAlsoCauseALineBreak")
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_single_field_struct() {
    let ast = struct_def("MyStruct").field("field", "i32").build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_nested_struct() {
    let ast = struct_def("Outer")
        .field("inner", "Inner")
        .field("another_field", "i32")
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_long_binary_expression() {
    let ast = fn_def("foo")
        .statement(expr().binary(
            expr().lit("a_very_long_string_that_should_cause_a_line_break"),
            BinOp::Add,
            expr().lit("another_very_long_string_that_should_also_cause_a_line_break"),
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_trait() {
    let ast = trait_def("MyTrait")
        .leading_comment(comment().line(" A simple trait."))
        .item(trait_item_fn("my_func"))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_loop_expression() {
    let ast = fn_def("foo")
        .statement(expr().loop_expr(block().statement(expr().lit(1)).build()))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_while_expression() {
    let ast = fn_def("foo")
        .statement(expr().while_loop(expr().lit(1), block().statement(expr().lit(2)).build()))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_for_expression() {
    let ast = fn_def("foo")
        .statement(expr().for_loop(
            pat().ident("x", false),
            expr().lit(1),
            block().statement(expr().lit(2)).build(),
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_assign_expression() {
    let ast = fn_def("foo")
        .statement(expr().assign(expr().lit("x"), expr().lit(1)))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_macro_call_expression() {
    let ast = fn_def("foo")
        .statement(expr().macro_call(
            "println",
            Delimiter::Parenthesis,
            thin_vec![TokenTree::Literal("hello".into())],
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_macro_call_expression_with_path() {
    let ast = fn_def("foo")
        .statement(expr().macro_call(
            path("std").segment("println").build(),
            Delimiter::Parenthesis,
            thin_vec![TokenTree::Literal("hello".into())],
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_enum() {
    let ast = enum_def("MyEnum")
        .leading_comment(comment().line(" A simple enum."))
        .variant("Variant1")
        .variant("Variant2")
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_impl() {
    let ast = impl_block("MyStruct")
        .function(fn_def("new").build())
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_trait_impl() {
    let ast = impl_block("MyStruct")
        .trait_("MyTrait")
        .function(fn_def("new").build())
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_unsafe_trait_impl() {
    let ast = impl_block("MyStruct")
        .trait_("MyTrait")
        .unsafe_()
        .function(fn_def("new").build())
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_negative_impl() {
    let ast = impl_block("MyStruct").trait_("MyTrait").negative().build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_let_statement() {
    let ast = fn_def("foo")
        .statement(
            stmt()
                .local(pat().ident("x", false))
                .ty("i32")
                .expr(expr().lit(42))
                .build(),
        )
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_if_expression() {
    let ast = fn_def("foo")
        .statement(expr().if_expr(
            expr().lit(1),
            block().statement(expr().lit(2)).build(),
            Some(expr().if_expr(
                expr().lit(3),
                block().statement(expr().lit(4)).build(),
                Some(expr().block(block().statement(expr().lit(5)).build())),
            )),
        ))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_binary_expression() {
    let ast = fn_def("foo")
        .statement(expr().binary(expr().lit(1), BinOp::Add, expr().lit(2)))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_expr_statement_without_semicolon() {
    let ast = fn_def("foo")
        .has_trailing_semicolon(false)
        .statement(expr().lit(42))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_item_statement() {
    let ast = fn_def("foo")
        .statement(stmt().item(struct_def("MyStruct").build()))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

use rasto::builder::path;

#[test]
fn test_macro_call_statement() {
    let ast = fn_def("foo")
        .statement(stmt().mac_call(ExprMacroCall {
            path: path("println").build(),
            delimiter: Delimiter::Parenthesis,
            tokens: thin_vec![TokenTree::Literal("hello".into())].into(),
        }))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_all_literals() {
    let ast = fn_def("literals")
        .statement(stmt().local("s").expr(expr().lit("hello")))
        .statement(stmt().local("bs").expr(expr().lit(b"hello")))
        .statement(stmt().local("cs").expr(expr().lit(c"hello")))
        .statement(stmt().local("b").expr(expr().lit(b'h')))
        .statement(stmt().local("c").expr(expr().lit('h')))
        .statement(stmt().local("i").expr(expr().lit(42)))
        .statement(
            stmt()
                .local("i_suffix")
                .expr(expr().int_lit_with_suffix(42, IntSuffix::U32)),
        )
        .statement(stmt().local("f").expr(expr().lit(1.23)))
        .statement(
            stmt()
                .local("f_suffix")
                .expr(expr().lit(Lit::Float(LitFloat::with_suffix("1.23", FloatSuffix::F32)))),
        )
        .statement(stmt().local("t").expr(expr().lit(true)))
        .build();

    insta::assert_snapshot!(pretty(&ast));
}

#[test]
fn test_struct() {
    let ast = struct_def("Foo")
        .leading_comment(comment().line(" A simple struct."))
        .field("field1", "i32")
        .field("field2", "String")
        .build();

    insta::assert_snapshot!(pretty(&ast));
}
