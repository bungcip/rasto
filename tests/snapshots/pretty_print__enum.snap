---
source: tests/pretty_print.rs
expression: pretty_print_item(ast)
---
// A simple enum.
enum MyEnum { Variant1, Variant2, }
